storage/range.go

```diff
@@ -1370,3 +1402,119 @@ func (r *Range) AdminSplit(args *proto.AdminSplitRequest, reply *proto.AdminSpli
 		reply.SetGoError(util.Errorf("split at key %q failed: %s", splitKey, err))
 	}
 }
+
+// InitialState implements the raft.Storage interface.
+func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
+	data, err := r.rm.Engine().Get(
+		proto.EncodedKey(engine.RaftStateKey(uint64(r.Desc.RaftID))))
+	if err != nil {
+		return raftpb.HardState{}, raftpb.ConfState{}, err
+	}
+	var hs raftpb.HardState
+	err = gogoproto.Unmarshal(data, &hs)
+	if err != nil {
+		return raftpb.HardState{}, raftpb.ConfState{}, err
+	}
+
+	// TODO(bdarnell): load ConfState if we have a previous snapshot.
+	return hs, raftpb.ConfState{}, nil
+}
+
+// Entries implements the raft.Storage interface.
+// TODO(bdarnell): consider caching for recent entries.
```

Comment:
Hopefully that's true, although I've been disappointed with sstable caching before. Log entries have very definite patterns (they're certain to be accessed several times shortly after being written, but after being committed they're unlikely to be touched), which might let us do better.